File: packages/agents/src/messaging/connection.py
"""RabbitMQ connection pool with retry logic and exponential backoff"""
import logging
import time
from typing import Optional
import pika
from pika.adapters.blocking_connection import BlockingConnection
from pika.exceptions import AMQPConnectionError, AMQPChannelError

from ..config import RabbitMQConfig

logger = logging.getLogger(__name__)


class ConnectionPool:
    """Manages RabbitMQ connections with retry logic and exponential backoff"""

    def __init__(
        self,
        config: RabbitMQConfig,
        max_retries: int = 5,
        initial_retry_delay: float = 1.0,
        max_retry_delay: float = 60.0,
    ):
        """
        Initialize connection pool

        Args:
            config: RabbitMQ configuration
            max_retries: Maximum number of connection retry attempts
            initial_retry_delay: Initial delay between retries in seconds
            max_retry_delay: Maximum delay between retries in seconds
        """
        self.config = config
        self.max_retries = max_retries
        self.initial_retry_delay = initial_retry_delay
        self.max_retry_delay = max_retry_delay
        self._connection: Optional[BlockingConnection] = None
        self._connection_params = self._build_connection_params()

    def _build_connection_params(self) -> pika.ConnectionParameters:
        """Build connection parameters from config"""
        credentials = pika.PlainCredentials(
            username=self.config.username,
            password=self.config.password,
        )

        return pika.ConnectionParameters(
            host=self.config.host,
            port=self.config.port,
            virtual_host=self.config.vhost,
            credentials=credentials,
            heartbeat=600,
            blocked_connection_timeout=300,
        )

    def _calculate_retry_delay(self, attempt: int) -> float:
        """
        Calculate exponential backoff delay

        Args:
            attempt: Current retry attempt number (0-indexed)

        Returns:
            Delay in seconds
        """
        delay = self.initial_retry_delay * (2**attempt)
        return min(delay, self.max_retry_delay)

    def connect(self) -> BlockingConnection:
        """
        Establish connection to RabbitMQ with retry logic

        Returns:
            Active RabbitMQ connection

        Raises:
            AMQPConnectionError: If connection fails after all retries
        """
        if self._connection and self._connection.is_open:
            return self._connection

        last_error: Optional[Exception] = None

        for attempt in range(self.max_retries):
            try:
                logger.info(
                    f"Attempting to connect to RabbitMQ at "
                    f"{self.config.host}:{self.config.port} (attempt {attempt + 1}/{self.max_retries})"
                )
                self._connection = pika.BlockingConnection(self._connection_params)
                logger.info("Successfully connected to RabbitMQ")
                return self._connection

            except (AMQPConnectionError, AMQPChannelError) as e:
                last_error = e
                if attempt < self.max_retries - 1:
                    delay = self._calculate_retry_delay(attempt)
                    logger.warning(
                        f"Failed to connect to RabbitMQ: {e}. "
                        f"Retrying in {delay:.2f} seconds..."
                    )
                    time.sleep(delay)
                else:
                    logger.error(
                        f"Failed to connect to RabbitMQ after {self.max_retries} attempts"
                    )

        raise AMQPConnectionError(
            f"Could not connect to RabbitMQ after {self.max_retries} attempts: {last_error}"
        )

    def get_connection(self) -> BlockingConnection:
        """
        Get active connection, reconnecting if necessary

        Returns:
            Active RabbitMQ connection
        """
        if not self._connection or not self._connection.is_open:
            return self.connect()
        return self._connection

    def close(self) -> None:
        """Close the connection gracefully"""
        if self._connection and self._connection.is_open:
            try:
                logger.info("Closing RabbitMQ connection")
                self._connection.close()
            except Exception as e:
                logger.error(f"Error closing RabbitMQ connection: {e}")
            finally:
                self._connection = None

    def is_connected(self) -> bool:
        """Check if connection is active"""
        return self._connection is not None and self._connection.is_open

    def __enter__(self) -> "ConnectionPool":
        """Context manager entry"""
        self.connect()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        """Context manager exit"""
        self.close()


# Global connection pool instance
_connection_pool: Optional[ConnectionPool] = None


def get_connection_pool(config: Optional[RabbitMQConfig] = None) -> ConnectionPool:
    """
    Get or create global connection pool instance

    Args:
        config: RabbitMQ configuration (required on first call)

    Returns:
        Global connection pool instance

    Raises:
        ValueError: If config is not provided on first call
    """
    global _connection_pool

    if _connection_pool is None:
        if config is None:
            raise ValueError("Config must be provided when creating connection pool")
        _connection_pool = ConnectionPool(config)

    return _connection_pool

File: packages/api/src/messaging/connection.ts
/**
 * RabbitMQ connection management with retry logic and exponential backoff
 */
import amqp, { Connection, Options } from 'amqplib';

export interface RabbitMQConfig {
  host: string;
  port: number;
  username: string;
  password: string;
  vhost: string;
}

export interface ConnectionPoolOptions {
  maxRetries?: number;
  initialRetryDelay?: number;
  maxRetryDelay?: number;
}

export class ConnectionPool {
  private config: RabbitMQConfig;
  private maxRetries: number;
  private initialRetryDelay: number;
  private maxRetryDelay: number;
  private connection: Connection | null = null;

  constructor(config: RabbitMQConfig, options: ConnectionPoolOptions = {}) {
    this.config = config;
    this.maxRetries = options.maxRetries ?? 5;
    this.initialRetryDelay = options.initialRetryDelay ?? 1000; // milliseconds
    this.maxRetryDelay = options.maxRetryDelay ?? 60000; // milliseconds
  }

  /**
   * Build connection URL from config
   */
  private buildConnectionUrl(): string {
    const { username, password, host, port, vhost } = this.config;
    const encodedVhost = encodeURIComponent(vhost);
    return `amqp://${username}:${password}@${host}:${port}/${encodedVhost}`;
  }

  /**
   * Calculate exponential backoff delay
   */
  private calculateRetryDelay(attempt: number): number {
    const delay = this.initialRetryDelay * Math.pow(2, attempt);
    return Math.min(delay, this.maxRetryDelay);
  }

  /**
   * Sleep for specified milliseconds
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Establish connection to RabbitMQ with retry logic
   */
  async connect(): Promise<Connection> {
    if (this.connection) {
      return this.connection;
    }

    const url = this.buildConnectionUrl();
    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        console.log(
          `Attempting to connect to RabbitMQ at ${this.config.host}:${this.config.port} ` +
            `(attempt ${attempt + 1}/${this.maxRetries})`
        );

        const options: Options.Connect = {
          heartbeat: 60,
        };

        this.connection = await amqp.connect(url, options);

        // Handle connection errors
        this.connection.on('error', (err) => {
          console.error('RabbitMQ connection error:', err);
          this.connection = null;
        });

        this.connection.on('close', () => {
          console.warn('RabbitMQ connection closed');
          this.connection = null;
        });

        console.log('Successfully connected to RabbitMQ');
        return this.connection;
      } catch (error) {
        lastError = error as Error;

        if (attempt < this.maxRetries - 1) {
          const delay = this.calculateRetryDelay(attempt);
          console.warn(
            `Failed to connect to RabbitMQ: ${lastError.message}. ` +
              `Retrying in ${delay}ms...`
          );
          await this.sleep(delay);
        } else {
          console.error(
            `Failed to connect to RabbitMQ after ${this.maxRetries} attempts`
          );
        }
      }
    }

    throw new Error(
      `Could not connect to RabbitMQ after ${this.maxRetries} attempts: ${lastError?.message}`
    );
  }

  /**
   * Get active connection, reconnecting if necessary
   */
  async getConnection(): Promise<Connection> {
    if (!this.connection) {
      return this.connect();
    }
    return this.connection;
  }

  /**
   * Close the connection gracefully
   */
  async close(): Promise<void> {
    if (this.connection) {
      try {
        console.log('Closing RabbitMQ connection');
        await this.connection.close();
      } catch (error) {
        console.error('Error closing RabbitMQ connection:', error);
      } finally {
        this.connection = null;
      }
    }
  }

  /**
   * Check if connection is active
   */
  isConnected(): boolean {
    return this.connection !== null;
  }
}

// Global connection pool instance
let connectionPool: ConnectionPool | null = null;

/**
 * Get or create global connection pool instance
 */
export function getConnectionPool(config?: RabbitMQConfig): ConnectionPool {
  if (!connectionPool) {
    if (!config) {
      throw new Error('Config must be provided when creating connection pool');
    }
    connectionPool = new ConnectionPool(config);
  }
  return connectionPool;
}
